libs/vm/src/analyser.cpp

```diff
@@ -357,25 +390,438 @@ bool Analyser::Analyse(BlockNodePtr const &root, std::vector<std::string> &error
     file_errors_array_.clear();
     return false;
   }
-
   errors.clear();
   return true;
 }
 
+void Analyser::EnforceLedgerRestrictions(BlockNodePtr const &block_node)
+{
+  LedgerRestrictionMetadata metadata;
+
+  ValidateBlock(block_node, metadata);
+
+  CheckInitFunctionUnique(metadata);
+  if (CheckSynergeticFunctionsPresentAndUnique(metadata))
+  {
+    CheckSynergeticContract(metadata);
+  }
+}
+
+void Analyser::ValidateBlock(BlockNodePtr const &block_node, LedgerRestrictionMetadata &metadata)
+{
+  for (NodePtr const &child : block_node->block_children)
+  {
+    switch (child->node_kind)
+    {
+    case NodeKind::File:
+    {
+      BlockNodePtr file_node = ConvertToBlockNodePtr(child);
+      filename_              = file_node->text;
+      ValidateBlock(file_node, metadata);
+      break;
+    }
+    case NodeKind::ContractDefinition:
+    case NodeKind::StructDefinition:
+    {
+      ExpressionNodePtr name_node = ConvertToExpressionNodePtr(child->children[0]);
+      TypePtr           type      = name_node->type;
+      if (type)
+      {
+        ValidateBlock(ConvertToBlockNodePtr(child), metadata);
+      }
+      break;
+    }
+    case NodeKind::ContractFunction:
+    {
+      ValidateFunctionAnnotations(child);
+      ValidateFunctionPrototype(child, metadata);
+      break;
+    }
+    case NodeKind::FreeFunctionDefinition:
+    case NodeKind::MemberFunctionDefinition:
+    {
+      ValidateFunctionAnnotations(ConvertToBlockNodePtr(child));
+      ValidateFunctionPrototype(ConvertToBlockNodePtr(child), metadata);
+      break;
+    }
+    default:
+    {
+      break;
+    }
+    }  // switch
+  }
+}
+
+void Analyser::ValidateFunctionAnnotations(NodePtr const &function_node)
+{
+  auto const &annotations_node = function_node->children[0];
+  if (annotations_node == nullptr)
+  {
+    return;
+  }
+  assert(annotations_node->node_kind == NodeKind::Annotations);
+
+  if (annotations_node->children.size() != 1u)
+  {
+    AddError(annotations_node->line, "Functions are only allowed to bear one annotation");
+  }
+
+  for (auto const &annotation : annotations_node->children)
+  {
+    auto const &text = annotation->text;
+    if (text != INIT_ANNOTATION && text != ACTION_ANNOTATION && text != QUERY_ANNOTATION &&
+        text != PROBLEM_ANNOTATION && text != OBJECTIVE_ANNOTATION && text != WORK_ANNOTATION &&
+        text != CLEAR_ANNOTATION)
+    {
+      AddError(annotation->line, "Invalid annotation");
```

Comment:
Maybe add the bad annotation to the error message?

Reply:
done